switch tests from constrain_as* to torch._check*

avikchaudhuri opened the pull request:
To fix data-dependent errors we want to recommend that people use `torch._check*` APIs. The `constrain_as*` APIs should be fully subsumed by them, and in the future we should kill them entirely.

Differential Revision: D56774333


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @ezyang @msaroufim @bdhirsh @anijain2305 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

avikchaudhuri commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -702,7 +702,8 @@ def tolist(self):
             s = shape_env.create_unbacked_symint()
             # max value?
-            torch._constrain_as_size(s, min=2)
+            torch._check_is_size(s)
```
Is this safe? Is the `>= 2` needed below?

angelayi replied:
I wonder if the >= 2 was added due to the 0/1 specialization issue. Also not sure if the 0/1 specialization issue has been fixed for this case. @ezyang

avikchaudhuri replied:
I think `_check_is_size` temporarily assumes `>= 2` whereas the check below will always force it. So I'm leaving in the check, don't want to change any behavior in this test right now.

angelayi commented on test/export/test_export.py:
```diff
@@ -3897,9 +3912,10 @@ def forward(self, x):
         class Foo(torch.nn.Module):
             def forward(self, x, y):
-                y = torch._constrain_as_size(y.item(), min=2)
+                y = torch.sym_constrain_range_for_size(y.item(), min=2)
```
Do you know if `torch.sym_constrain_range_for_size` is a maintained API? I sort of forgot this API existed 😅 . I thought we wanted to move everything to `torch._check` / `torch._check_is_size`

avikchaudhuri replied:
Not sure. Leaving the TODO for now, yeah I don't want to recommend it, but it exists and I don't know why. All I know is that `_constrain_as*` directly desugars into this.

tugsbayasgalan replied:
Ahh i think it is ok to delete it. It was for preserving the op in the aten graph so we can always end up re-constraining when we retrace.

angelayi commented on test/test_proxy_tensor.py:
```diff
@@ -1498,6 +1498,7 @@ def f(lengths, values):
             # tolist not directly supported atm
             sizes = [lengths[i].item() for i in range(lengths.size(0))]
             for s in sizes:
+                # TODO(avik): no assertion generated with torch._check_is_size?
```
Is this resulting in an error? It seems like this is because `torch._check_is_size` is a softer version of `_constrain_as_size`, where it will only add assertions if we cannot prove something at compile-time. https://fburl.com/code/sm5ojgw2

avikchaudhuri replied:
Oh interesting, deserves a follow-up.

avikchaudhuri commented on test/export/test_export.py:
```diff
@@ -2152,11 +2153,12 @@ def forward(self, x):
 
-    def test_constrain_value_with_no_default(self):
+    def test_sym_constrain_range_with_no_default(self):
+        # TODO(avik): update this test with torch._check
```
OK, will delete.

avikchaudhuri committed:
switch tests from constrain_as* to torch._check*

Summary: To fix data-dependent errors we want to recommend that people use `torch._check*` APIs. The `constrain_as*` APIs should be fully subsumed by them, and in the future we should kill them entirely.

Test Plan: updated tests

Reviewed By: tugsbayasgalan

Differential Revision: D56774333

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -2262,17 +2229,19 @@ def forward(self, x, y):
         self.assertTrue(torch.allclose(ep.module()(*test_inp), fn(*test_inp)))
 
     @testing.expectedFailureNonStrict
+    @testing.expectedFailureRetraceability
```
Could you link the task that you filed for this?

avikchaudhuri replied:
Might be related to other tasks, there's an error message diff:
> AssertionError: "is outside of inline constraint" does not match "Runtime assertion failed for _local_scalar_dense_default >= 2"

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -2476,11 +2464,13 @@ def forward(self, x):
         ) as cm:
             ep.module()(torch.tensor([30]))
 
+    @testing.expectedFailureNonStrict  # assert not found
```
Don't we have the deferred runtime assertion now in both non-strict and strict?

avikchaudhuri replied:
Not sure.
> RuntimeError: Expected to find "torch.ops.aten._assert_async.msg" but did not find it

tugsbayasgalan replied:
cc @BoyuanFeng
